libvast/src/system/node.cpp

```diff
@@ -604,37 +603,38 @@ node(node_actor::stateful_pointer<node_state> self, std::string name,
                    msg.source);
       }
     });
-    std::vector<caf::actor> teardown_aux;
-    for (const auto& [_, comp] : self->state.registry.components()) {
+    auto& registry = self->state.registry;
+    // Core components are terminated in a second stage, we remove them from the
+    // registry upfront and deal with them later.
+    std::vector<caf::actor> core_components;
+    // The components listed here need to be terminated in sequential order.
+    // The importer needs to shut down first because it might still have
+    // buffered data. The index uses the archive for querying. The filesystem
+    // is needed by all others for the persisting logic.
+    for (const char* name : {"importer", "index", "archive", "filesystem"}) {
+      if (auto comp = registry.remove(name))
+        core_components.push_back(comp->actor);
```

Comment:
This is missing a `demonitor` call. Also the `else` should probably print an error, or we should even assert here.

Reply:
The `demonitor` happens in `core_shutdown_sequence`.
I didn't not include an `else` branch because the exit handler can also be invoked because one of the core components went down for some reason.

Reply:
Sane reasoning.